{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "input_dir = '/home/petadimensionlab/Desktop/PRJNA320132'\n",
    "output_dir =  '/home/petadimensionlab/Desktop/PRJNA320132'\n",
    "## reference file ##\n",
    "gg97fasta = '/home/petadimensionlab/ws/ref/gg_13_5_otus/rep_set/97_otus.fasta'\n",
    "gg97tax = '/home/petadimensionlab/ws/ref/gg_13_5_otus/taxonomy/97_otu_taxonomy.txt'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os, glob\n",
    "#### configurations ####\n",
    "is_taxonomy = 'yes'\n",
    "is_denovopicking = 'yes'\n",
    "is_mapping = 'yes'\n",
    "is_deletefiles = 'yes'\n",
    "## softwares ##\n",
    "vsearch = '/home/petadimensionlab/miniconda3/pkgs/vsearch-2.7.0-1/bin/vsearch'\n",
    "## input / output file ##\n",
    "config_file_name = 'config.csv'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "## parameters ##\n",
    "maxdiffs = 5 #10\n",
    "minmergelen = 1\n",
    "maxmergelen = 600\n",
    "minseqlength = 50\n",
    "maxee_rate = 0.5/100 #%\n",
    "minsize = 2\n",
    "thr = 0.97\n",
    "threadsnum = 4\n",
    "## file names ##\n",
    "filtered_file_name = 'filtered.fasta'\n",
    "centroids_file_name = 'centroids.fasta'\n",
    "denovootu_file_name = 'denovootu.txt'\n",
    "hits_file_name = 'hits.txt'\n",
    "ggotuname_file_name = 'ggotu_name.txt'\n",
    "ggotuID_file_name = 'ggotu.txt'\n",
    "\n",
    "## filtering ##\n",
    "filtered_file = os.path.join(output_dir,filtered_file_name)\n",
    "centroids_file = os.path.join(output_dir,centroids_file_name)\n",
    "\n",
    "## Assign taxonomy ##\n",
    "hits_file = os.path.join(output_dir,hits_file_name)\n",
    "if is_taxonomy=='yes':\n",
    "    cmd = '%s --usearch_global %s --db %s --strand both --id %s --threads %s --userfields \\\"query+target+id\\\" --userout %s' % (vsearch,centroids_file,gg97fasta,str(thr),str(threadsnum),hits_file)\n",
    "    os.system(cmd)\n",
    "else:\n",
    "    msg = 'Skip taxonomy assignment.'\n",
    "    print( msg )\n",
    "\n",
    "## de novo picking ##\n",
    "denovootu_file = os.path.join(output_dir,denovootu_file_name)\n",
    "if is_denovopicking=='yes':\n",
    "    cmd = '%s --usearch_global %s --db %s --strand both --id %s --threads %s --otutabout %s' % (vsearch,filtered_file,centroids_file,str(thr),str(threadsnum),denovootu_file)\n",
    "    os.system(cmd)\n",
    "else:\n",
    "    msg = 'Skip denovo picking.'\n",
    "    print( msg )\n",
    "\n",
    "## Mapping to GreenGene database ##\n",
    "if is_mapping=='yes':\n",
    "    ggid2ggname = {}\n",
    "    frgg97 = open(gg97tax,'r').readlines()\n",
    "    for line in frgg97:\n",
    "        line = line.replace('\\n','')\n",
    "        tmp = line.split('\\t')\n",
    "        ggid2ggname[tmp[0]] = tmp[1]\n",
    "    ## make a mapping from \"user defined sequence ID\" to \"greengene ID\"\n",
    "    myid2ggid = {}\n",
    "    myid2ggname = {}\n",
    "    frh = open(hits_file,'r').readlines()\n",
    "    for line in frh:\n",
    "        tmp = line.split('\\t') # ggID: tmp[1]\n",
    "        ggid = tmp[1]\n",
    "        ggname = ggid2ggname[ggid]\n",
    "        myids = tmp[0].split(';') # myID : tmp[0]\n",
    "        myid2ggid[myids[0]] = ggid\n",
    "        myid2ggname[myids[0]] = ggname\n",
    "\n",
    "    frd = open(denovootu_file,'r').readlines()\n",
    "    header = frd[0] # restore header information\n",
    "    frd.pop(0) # remove the header line\n",
    "    ## taxonomy name version ##\n",
    "    ggotu_name_file = os.path.join(output_dir,ggotuname_file_name)\n",
    "    fw = open(ggotu_name_file,'w')\n",
    "    fw.write(header)\n",
    "    for line in frd:\n",
    "        tmp = line.split('\\t') # tmp[0]: myID\n",
    "        try:\n",
    "            tmp[0] = myid2ggname[tmp[0]]\n",
    "            fw.write('\\t'.join(tmp))\n",
    "        except:\n",
    "            continue\n",
    "    fw.close()\n",
    "    ## taxonomy ID version ##\n",
    "    ggotu_ID_file = os.path.join(output_dir,ggotuID_file_name)\n",
    "    fw = open(ggotu_ID_file,'w')\n",
    "    fw.write(header)\n",
    "    for line in frd:\n",
    "        tmp = line.split('\\t') # tmp[0]: myID\n",
    "        try:\n",
    "            tmp[0] = myid2ggid[tmp[0]]\n",
    "            fw.write('\\t'.join(tmp))\n",
    "        except:\n",
    "            continue\n",
    "    fw.close()\n",
    "else:\n",
    "    msg = 'Skip taxonomy mapping.'\n",
    "    print( msg )\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "## delete intermediate files ##\n",
    "if is_deletefiles=='yes':\n",
    "    paired_files = glob.glob('paired*.fastq.gz')\n",
    "    unpaired_files = glob.glob('unpaired*.fastq.gz')\n",
    "    merged_files = glob.glob('*_merged.fasta')\n",
    "    rm_files = ['appended.fastq','filtered.fasta','unique.fasta']+paired_files+unpaired_files+merged_files\n",
    "    for file in rm_files:\n",
    "        cmd = 'rm %s' % (file)\n",
    "        os.system(cmd)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.5.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
